gemd/entity/object/base_object.py

```diff
@@ -52,19 +60,52 @@ def _attribute_has_setter(cls, name: str) -> bool:
         return prop is None or prop.fset is not None
 
     @property
-    def name(self):
+    def name(self) -> str:
         """Get name."""
         return self._name
 
     @name.setter
-    def name(self, name):
+    def name(self, name: str):
         self._name = validate_str(name)
 
     @property
-    def file_links(self):
+    def file_links(self) -> List[FileLink]:
         """Get file links."""
         return self._file_links
 
     @file_links.setter
-    def file_links(self, file_links):
+    def file_links(self, file_links: Union[Iterable[FileLink], FileLink]):
         self._file_links = validate_list(file_links, FileLink)
+
+    def all_dependencies(self) -> Set[BaseEntity]:
+        """Return a set of all immediate dependencies (no recursion)."""
+        from gemd.entity.object.has_parameters import HasParameters
+        from gemd.entity.object.has_conditions import HasConditions
+        from gemd.entity.object.has_properties import HasProperties
+        from gemd.entity.object.has_spec import HasSpec
+        from gemd.entity.object.has_template import HasTemplate
+
+        from gemd.entity.object import IngredientRun, MaterialRun, MeasurementRun  # no ProcessRun
+        from gemd.entity.object import IngredientSpec, MaterialSpec  # no ProcessSpec
+
+        result = set()
+
+        for typ in (HasParameters, HasConditions, HasProperties, HasSpec, HasTemplate):
+            if isinstance(self, typ):
+                result |= typ.all_dependencies(self)
+        if isinstance(self, MaterialSpec):  # is structured inconsistently
+            for attr in self.properties:
+                if attr.property.template is not None:
+                    result.add(attr.property.template)
+                for condition in attr.conditions:
+                    if condition.template is not None:
+                        result.add(condition.template)
+
+        if isinstance(self, (IngredientRun, IngredientSpec, MeasurementRun)):
+            if self.material is not None:
+                result.add(self.material)
+        if isinstance(self, (IngredientRun, IngredientSpec, MaterialRun, MaterialSpec)):
+            if self.process is not None:
+                result.add(self.process)
+
+        return result
```

Comment:
This logic feels wrong because it requires `base_object` to know about the structure of its child classes.

Reply:
Yeah, I agree that the `isinstance(self, GEMD)` would look nicer if we could test mixins instead.  No obvious solution for `MaterialSpec` since `PropertyAndConditions` is just structured differently than `MeasurementRun`'s property.  A `HasMaterial` and `HasProcess` mixin would be more straightforward, but it's not clear that it buys you much because the only consequence is that you provide an abstract method that must be implemented where you don't even know the return type.

If necessary, I could bake that chunk of framework out.

Reply:
If we can't write this method without having a holistic understanding of the GEMD object structure, then what's the value in having this logic live in `BaseObject` as opposed to citrine-python?

Reply:
Especially since `all_dependencies` is only used for batching in the context of upload, which is a citrine-python concern

Reply:
The mechanism has been reworked.  Does this make more sense?  I generally don't think that logic that relies only on gemd-python concepts should be implemented in a downstream package, and it seemed like implementing it as a `gemd.util` method was going to be messy.  I still would have liked a cleaner way of accessing the PropertyAndCondition elements.

Reply:
Yes, I think that this makes more sense.